packages/nimble-components/src/select/specs/filterable-select-hld.md

```diff
@@ -20,29 +20,58 @@ Our clients have a need of a filterable dropdown component that does not allow a
 
 #### Select
 
-We will provide a means for clients to enable this feature with a new attribute:
+We will provide a means for clients to enable this feature with some new attributes:
 
 ```ts
 export class Select() {
     ...
     @attr({ attribute: 'filter-mode' })
     public filterMode = FilterMode.none;
+
+    /**
+     * Displays a message at the bottom of the dropdown to indicate more options
+     * are currently being loaded.
+     */
+    @attr({ attribute: 'is-loading' })
+    public isLoading = false;
     ...
 }
 
 // Provide enum for filterMode to allow for future modes
 export const FilterMode  = {
     none: undefined;
     standard: 'standard';
+    custom: 'custom';
 } as const;
 ```
 
 -   The `standard` filterMode will result in case-insensitive, diacritic-insensitive filtering.
--   `filterMode` will default to `none` so as not to affect existing clients.
+-   The `custom` filterMode will display a search input, but it won't automatically filter the options as the user types. Instead, clients can use the `filter-input` event described in the ['Dynamic Options'](#dynamic-options) section below to implement their own filtering logic.
 -   The `none` filter mode results in no search input being shown in the dropdown.
+-   `filterMode` will default to `none` so as not to affect existing clients.
 
 _Note: The `filterMode` isn't meant to mirror the `Combobox` `autocomplete` API, as they do serve slightly different purposes: The `autocomplete` for the `Combobox` ultimately helps set the actual value of the `Combobox` as the user types, and isn't necessarily performing any filtering (e.g. the `inline` mode). One possible concern, however, is that we are presenting an API that will allow different types of filter behaviors (i.e. case sensitive) that the `Combobox` does not support. Additionally, I am proposing diacritic insensitive filtering, which the `Combobox` also does not currently support, but I feel this is quite likely a better default experience._
 
+#### Dynamic options
+
+The `filterMode: custom` option allows clients to dynamically supply list options to the `select` control in response to user search text, by subscribing to the `filter-input` event.
```

Comment:
Should we fire the event when the user first opens the input? This could be useful to trigger an initial network request for items. I could also imagine apps fetching the initial set when the page loads but that's not guaranteed.

Should we fire the event when the user types in the filter field if the filter mode is `standard`? I don't see a need but it'd be good to clarify.

Should Nimble debounce the event (I'm guessing not) or provide guidance about apps debouncing it?

Reply:
> Should we fire the event when the user first opens the input?

It would also be good to clarify what happens when the filter input is automatically cleared because the select closes. Would that use the same `filter-input` event? Or would that be a different event that fires when the select opens and/or closes?

Reply:
Here are my initial takes:
- My initial guess is that while this event could be leveraged, I don't think the use-case you outlined is the expected pattern. Waiting for the initial opening of the dropdown before populating the `Select` with an initial set of options doesn't seem typical, and it just makes the implementation more complicated in having to disambiguate the _first_ time the dropdown is opened versus the other times (I'm asserting that making a network request every time the dropdown opens to retrieve the same set of options is an anti-pattern). If clients end up finding a need for this I say we address it as a follow-up.
- At the moment, I was intending to fire the event even when the filter mode was set to `standard`. I'm willing to alter this decision however.
- Nimble should _not_ do any debouncing, in my opinion, in that there really isn't anything to debounce. We don't want to artificially delay the event (i.e. wait 200 ms for another key press before firing the even), as this could just lead to an unnecessary delay. Debouncing actions such as these is pretty standard procedure in a framework like Angular when the intent is to make a network request based on a user typing. Let's just expect clients to continue to use standard procedure here.

Reply:
Copying Molly's comment for discussion:
> Should we fire the event when the user first opens the input?

It would also be good to clarify what happens when the filter input is automatically cleared because the select closes. Would that use the same `filter-input` event? Or would that be a different event that fires when the select opens and/or closes?

Reply:
> It would also be good to clarify what happens when the filter input is automatically cleared because the select closes. Would that use the same `filter-input` event? Or would that be a different event that fires when the select opens and/or closes?

This is a great question. Would it be better to provide guidance to utilize RxJS patterns that will produce an observable set of options from an observable of options matching a current non-empty filter state, or if the filtered set is empty, then an observable that represents the "initial set" (but would be modified by selected options that weren't in initial set).

Reply:
Talked it through with @mollykreis, and I think we're in agreement that firing the `filter-input` when the dropdown closes (containing an empty string for the filter text) is sufficient for clients to achieve what they need (and ultimately this is just an implementation detail as clients should already be handling what should happen when the filter input is empty).

Reply:
Updated: I've added notes regarding debouncing, as well as when the `filter-input` event will be emitted.